block/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -72,16 +83,67 @@ retrievalWorker keepAliveTimer = worker outs (retrievalWorkerImpl keepAliveTimer
 --
 -- If both happen at the same time, 'BlockRetrievalQueue' takes precedence.
 --
+-- Note: if we haven't caught up with the blockchain yet, the worker will
+-- try to download and apply blockchain dumps, and only if it fails it'll
+-- switch to requesting headers/blocks.
+--
 retrievalWorkerImpl
     :: forall ctx m.
        (BlockWorkMode ctx m)
-    => Timer -> SendActions m -> m ()
-retrievalWorkerImpl keepAliveTimer SendActions {..} =
-    handleAll mainLoopE $ do
-        logInfo "Starting retrievalWorker loop"
-        mainLoop
+    => Timer -> Maybe Text -> SendActions m -> m ()
+retrievalWorkerImpl keepAliveTimer blockStorageMirror SendActions {..} = do
+    handleAll tryDownloadE $ do
+        epoch <- view epochIndexL <$> DB.getTipHeader
+        whenJust blockStorageMirror $ \dumpUrl ->
+            tryDownload epoch (toString dumpUrl)
+    handleAll recoveryLoopE $ do
+        logInfo "Starting recovery loop"
+        recoveryLoop
   where
-    mainLoop = do
+    -- | Try to download and apply a blockchain dump, starting from the
+    -- given epoch.
+    tryDownload epoch dumpUrl = whenNothingM_ getCurrentSlot $ do
+        let epochUrl = dumpUrl <//> format "epoch{}.cbor.lzma"
+                                           (toInteger epoch)
+        logInfo ("Downloading blockchain dump for epoch "+|epoch|+
+                 " from "+|epochUrl|+"")
+        isSuccess <- modifyStateLock HighPriority "tryDownload" $ \tip -> do
+            slot <- getEpochOrSlot <$> DB.getTipHeader
+            -- When we're already in the epoch we're going to be downloading
+            -- blocks from, it's possible that we're in slot 10000 or
+            -- something and we'll have to skip 10000 blocks before we can
+            -- show any progress. In this case let's warn about it.
+            whenRight (unEpochOrSlot slot) $ \slotId ->
+                when (slotId ^. epochIndexL == epoch) $
+                    logDebug ("Going to download blocks and skip until slot "
+                              +|view siSlotL slotId|+", this may take " <>
+                              "some time")
+            let applyDump
+                    = decodeBlockDumpC
+                   .| (C.dropWhileC ((/= tip) . view prevBlockL)
```

Comment:
Is it possible to assert here that not all blocks were `drop`ped?

Reply:
It can in theory happen if we're at the penultimate block in epoch N and the last block wasn't generated – in which case there are no blocks left to download in this epoch, but we can't know that and so we'll have to download and discard all blocks from epoch N before moving to the next epoch.

Reply:
Good point.